src/server.c

```diff
@@ -201,6 +201,32 @@ mstime_t mstime(void) {
     return ustime()/1000;
 }
 
+/* Return the command time snapshot in milliseconds.
+ * The time the command started is the logical time it runs,
+ * and all the time readings during the execution time should
+ * reflect the same time.
+ * More details can be found in the comments below. */
+mstime_t commandTimeSnapshot(void) {
+    /* If we are in the context of a Lua script, we pretend that time is
+     * blocked to when the Lua script started. This way a key can expire
+     * only the first time it is accessed and not in the middle of the
+     * script execution, making propagation to slaves / AOF consistent.
+     * See issue #1525 on Github for more information. */
+    if (server.script_caller) {
+        return scriptTimeSnapshot();
+    }
+    /* If we are in the middle of a command execution, we still want to use
+     * a reference time that does not change: in that case we just use the
+     * cached time, that we update before each call in the call() function.
+     * This way we avoid that commands such as RPOPLPUSH or similar, that
+     * may re-open the same key multiple times, can invalidate an already
+     * open object in a next call, if the next call will see the key expired,
+     * while the first did not. */
+    else {
+        return server.mstime;
```

Comment:
The new test added in #11176 failed because:
in getNodeByQuery, the lookupKey, here it used to use server.mstime(), and the new code we return server.mstime,
then produce different expired results.

Reply:
ohh, i missed this comment.
well, i see many ways to resolve it, listing them from worse to best.

1. add a PING in the [test](https://github.com/redis/redis/pull/11176/files#diff-8828084a4a321c4785aa317b418e6c440d4aaf46e41a2b23674d800e5a676882) between the sleep and the GET command. - this is obviously the wrong fix!
2. revert the change in f80cb16a67240ba30d026a9d69423feae368c04d and revive the ability of commandTimeSnapshot to sample the actual `mstime()` when called outside the context of `call()`. - this leaves hidden / excessive calls to `mstime()` on each command executed in cluster mode.
3. add an additional call to `updateCachedTime` in processCommand. - this will slow things down, see #10502
4. somehow move the call for `updateCachedTime` from `call` upwards to `processCommand`. - this will mess up the time measurements for commandstats, etc.
5. move the call for `updateCachedTime` from `serverCron` to `afterSleep`. effectively getting the same outcome of the first proposed solution above.

@guybe7 @soloestoy WDYT?